src/tactic/itauto.lean

```diff
@@ -0,0 +1,486 @@
+      end⟩
+    end
+  end
+
+/-- The main prover. This receives a context of proven or assumed lemmas and a target proposition,
+and returns a proof or `none` (with state for the fresh variable generator).
+The intuitionistic logic rules are separated into three groups:
+
+* level 1: No splitting, validity preserving: apply whenever you can.
+  Left rules in `context.add`, right rules in `prove`
+* level 2: Splitting rules, validity preserving: apply after level 1 rules. Done in `prove`
+* level 3: Splitting rules, not validity preserving: apply only if nothing else applies.
+  Done in `search`
+
+The level 1 rules on the right of the turnstile are `Γ ⊢ ⊤` and `Γ ⊢ A → B`, these are easy to
+handle. The rule `Γ ⊢ A ∧ B` is a level 2 rule, also handled here. If none of these apply, we try
+the level 2 rule `A ∨ B ⊢ C` by searching the context and splitting all ors we find. Finally, if
+we don't make any more progress, we go to the search phase.
+-/
+meta def prove : context → prop → state_t ℕ option proof
+| Γ prop.true := pure proof.triv
+| Γ (prop.imp A B) := do
+  a ← fresh_name,
+  proof.intro a A <$> Γ.with_add A (proof.hyp a) B prove
+| Γ (prop.and ak A B) := do
+  let (A, B) := ak.sides A B,
+  p ← prove Γ A,
+  q ← prove Γ B,
+  pure (p.and_intro ak q)
+| Γ B := Γ.fold (search prove B) (λ A p IH Γ,
+    match A with
+    | prop.or A₁ A₂ := do
+      let Γ : context := Γ.erase A,
+      a ← fresh_name,
+      p₁ ← Γ.with_add A₁ (proof.hyp a) B (λ Γ _, IH Γ),
+      p₂ ← Γ.with_add A₂ (proof.hyp a) B (λ Γ _, IH Γ),
+      pure (proof.or_elim A₁ A₂ p a p₁ p₂)
+    | _ := IH Γ
+    end) Γ
+
+/-- Reifies an atomic or otherwise unrecognized proposition. If it is defeq to a proposition we
+have already allocated, we reuse it, otherwise we name it with a new index. -/
+meta def reify_atom (atoms : ref (buffer expr)) (e : expr) : tactic prop := do
+  vec ← read_ref atoms,
+  o ← try_core $ vec.iterate failure (λ i e' r,
+    r <|> (is_def_eq e e' >> pure i.1)),
+  match o with
+  | none := write_ref atoms (vec.push_back e) $> prop.var vec.size
+  | some i := pure $ prop.var i
+  end
+
+/-- Reify an `expr` into a `prop`, allocating anything non-propositional as an atom in the
+`atoms` list. -/
+meta def reify (atoms : ref (buffer expr)) : expr → tactic prop
+| `(true) := pure prop.true
+| `(false) := pure prop.false
+| `(¬ %%a) := flip prop.imp prop.false <$> reify a
+| `(%%a ∧ %%b) := prop.and and_kind.and <$> reify a <*> reify b
+| `(%%a ∨ %%b) := prop.or <$> reify a <*> reify b
+| `(%%a ↔ %%b) := prop.and and_kind.iff <$> reify a <*> reify b
+| `(@eq Prop %%a %%b) := prop.and and_kind.eq <$> reify a <*> reify b
+| `(@ne Prop %%a %%b) := do a ← reify a, b ← reify b, pure $ (a.and and_kind.eq b).imp prop.false
+| e@`(%%a → %%b) :=
+  if b.has_var then reify_atom atoms e else prop.imp <$> reify a <*> reify b
+| e := reify_atom atoms e
+
+/-- Once we have a proof object, we have to apply it to the goal. (Some of these cases are a bit
```

Comment:
It could be easier here to just construct the proof terms by hand (and maybe merge the function with proof.check).

Reply:
Faster: yes, easier: no. This style of construction is not my favorite but Jeremy warned me that since intuitionistic logic isn't particularly useful I should try not to spend too much time on optimizations. Let's file this one under future work.